chore(docs): adjust K8s documentation

georglauterbach opened the pull request:
# Description

<!-- Include a summary of the change.
     Please also include relevant motivation and context. -->

This PR updates the Kubernetes documentation, making it more straight-forward, easier to understand and more structured.

<!-- Link the issue which will be fixed (if any) here: -->
Closes #2094

## Type of change

<!-- Delete options that are not relevant. -->

- [x] Improvement (non-breaking change that does improve existing functionality)
- [x] This change **is** a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

polarathene commented on docs/content/config/advanced/kubernetes.md:
```diff
@@ -460,79 +432,18 @@ Then, configure both [Postfix][docs-postfix] and [Dovecot][docs-dovecot] to expe
-- Not possible to access mailserver via inner cluster Kubernetes DNS, as PROXY protocol is required for incoming connections.
-
-## Let's Encrypt Certificates
+With this approach,
 
-[Kube-Lego][kube-lego] may be used for a role of Let's Encrypt client. It works with Kubernetes [Ingress Resources][k8s-network-ingress] and automatically issues/manages certificates/keys for exposed services via Ingresses.
```
Did this section lose relevance?

georglauterbach replied:
See my comment down below. Not sure why one would use Kube-Lego, _which is deprecated btw_, instead of certManager.

polarathene commented on docs/content/config/advanced/kubernetes.md:
````diff
@@ -460,79 +432,18 @@ Then, configure both [Postfix][docs-postfix] and [Dovecot][docs-dovecot] to expe
-        secret:
-          secretName: mailserver.tls
-    ```
+- is is not possible to access the mail server via cluster-DNS, as the PROXY protocol is required for incoming connections
````
```suggestion
- it is not possible to access the mail server via cluster-DNS, as the PROXY protocol is required for incoming connections
```

polarathene reviewed (changes requested):
- [Current live docs k8s page](https://docker-mailserver.github.io/docker-mailserver/edge/config/advanced/kubernetes/)
- [Current PR docs k8s page](https://pullrequest-2111--dms-doc-previews.netlify.app/config/advanced/kubernetes/)

---

Primarily minor corrections or phrasing suggestions.

- I would like to know why the last section was dropped. Other parts have been removed with no explanation either, such as the admonitions with notes/warnings in the current live docs.
- Any reason for dropping the usage of example admonitions, especially collapsible ones? Only one remains at the end.
- I understand this is WIP, some content needs to be broken apart into smaller paragraphs.
- I would like to have feedback from another k8s user/contributor, but presently the current docs seem more approachable.
- I'm not able to review the yaml configs, I trust you've done well with that part :)

georglauterbach commented on docs/content/config/advanced/kubernetes.md:
```diff
@@ -2,392 +2,363 @@
+
+## Exposing your Mail Server to the Outside World
+
+The more difficult part with K8s is to expose deployed mailserver to outside world. K8s provides multiple ways for doing that; each has its downsides and complexity. The major problem with exposing the mail server to outside world in K8s is to [preserve the real client IP][k8s-service-source-ip]. The real client IP is required by the mail server for performing IP-based SPF checks and spam checks. ...
```
Yes :D

georglauterbach commented on docs/content/config/advanced/kubernetes.md:
```diff
@@ -2,392 +2,363 @@
+With this approach,
 
-- Not possible to access mailserver via other cluster Nodes, only via the one mailserver deployed at.
-- Every Port within the Container is exposed on the Host side, regardless of what the `ports` section in the Configuration defines. 
+- it is not possible to access mailserver via other cluster Nodes, only via the one mailserver deployed at
+- every Port within the Container is exposed on the Host side
```
No, you should read the sentences as

"With this approach, it is not possible to access mailserver via other cluster Nodes, only via the one mailserver deployed at"
"With this approach, every Port within the Container is exposed on the Host side"

> Is this a formatting preference to use lower case list items instead of capitalizing the first letter of a sentence? Likewise with the removal of full stops at the end.

I removed the full stops by accident.

> The common subheading/subsection **Downsides** is replaced with a generic and less consistent message "This approach" or similar, even though the focus is on clearly identifying the downsides to an approach?

Using `**` does not indicate a subheading in Markdown (Markdownlint will even report this). I think that this section clarifies the consequences of using the approach, which are not always downsides.

polarathene replied:
> No, you should read the sentences as

Yeah, I still do when presented with a list. Personal bias of mine perhaps that the list items still are formatted like new sentences starting with capital letters. I don't mean to be too fussy about it :)

It's a little weird at the end of the doc with a single list item. Would a numbered list work better? Up to you, leave it if you think it's fine as-is.

> Using ** does not indicate a subheading in Markdown (Markdownlint will even report this).

Yes, the issue with actual heading is it's invalid for them to all have the same heading. At the time I didn't know how to resolve that, but I'm later learned of syntax to provide unique heading ids if we wanted the ToC and links to headings to work correctly.

> which are not always downsides.

Fair enough :+1:

georglauterbach commented on docs/content/config/advanced/kubernetes.md:
````diff
@@ -2,392 +2,363 @@
+```
 
-**Downsides**
+With this approach,
````
I feel like, with the idea I had of making the lists "sentences", I feel like the ":" does not fit here.

georglauterbach commented:
> I would like to know why the last section was dropped. Other parts have been removed with no explanation either, such as the admonitions with notes/warnings in the current live docs.

I removed the last section because on the one hand, we already have a section about certificates, on the other hand, if you want to have a dedicated section about certificates, why just write about Let'sEncrypt? I feel like with K8s, there is more to it, especially when you consider [certManager](https://cert-manager.io/docs/).

> Any reason for dropping the usage of example admonitions, especially collapsible ones? Only one remains at the end.

I found it odd that you'd need to unfold them, because this is the content everyone is there for...

> some content needs to be broken apart into smaller paragraphs.

Why do you figure? I must say I disagree.

georglauterbach commented on docs/content/config/advanced/kubernetes.md:
```diff
@@ -2,392 +2,363 @@
 ### Bind to concrete Node and use host network
 
-The simplest way to preserve real client IP is to use `hostPort` and `hostNetwork: true` in the mailserver [Pod][k8s-workload-pod]. This comes in price of availability: you can talk to mailserver from outside world only via IPs of [Node][k8s-nodes] where mailserver is deployed.
+One way to preserve the real client IP is to use `hostPort` and `hostNetwork: true`. This comes in price of availability: you can talk to the mail server from outside world only via IPs of [Node][k8s-nodes] where mailserver is deployed.
```
Just an inconsistency on my side. Will correct.

georglauterbach commented:
> I would like to have feedback from another k8s user/contributor, but presently the current docs seem more approachable.

I think that with the current docs, before you're able to work with it, you'd need to expand 4 admonitions, and you do not have a clear structure. But this is just my PoV.

polarathene commented:
> you'd need to expand 4 admonitions

I am not familiar with the yaml configs being shared, if all the snippets are relevant to the user you can collapse under 1 admonition for example with each clearly labeled, or if this information will always be relevant to the reader of this page, avoid any collapsing.

When there is optional configs, it often made more sense imo to reduce visual noise by collapsing it by default, a reader would scan through the document or ToC sidebar and get the information they want.

I can understand if that workflow differs for others. I personally prefer heavy pages to not be as overwhelming to someone seeking guidance by using progressive disclosure UX (_a few extra clicks at most vs overloading someone with everything all at once which usually isn't as well received, especially if only a portion is actually relevant to focus on_).

---

If the intention is to copy these configs as files, we can "import" them into the docs, they'll be rendered as they do currently, but we can additionally link to them on github to copy/paste from a clone of the repo. I believe I have instructions for that on my WIP documentation of the doc features (progress stalled due to other priorities). Would that be helpful?

---

> you do not have a clear structure

At least with the collapsed snippets, for someone familiar with k8s which seems to be the expectation here, the structure of them seemed clear that they were k8s related config and the names matched the `kind` metadata?

Your updates are helpful with the added context between them, but I think we could improve that a little, perhaps with some subheadings and not collapsing the description for each part into a single hefty paragraph? They're minor improvements compared to the value of your contribution, but they help make the content more easier to follow :sweat_smile: 

---

I removed the last section because on the one hand, we already have a section about certificates, on the other hand, if you want to have a dedicated section about certificates, why just write about Let'sEncrypt?

I don't mind it being removed, so long as we're not losing any value it may provide. If that information isn't of value or is already covered elsewhere no worries, otherwise we should perhaps migrate it somewhere more appropriate.

LetsEncrypt is popular enough that I don't mind it if it's helpful/relevant (we don't cover k8s anywhere else but this page right?). Removing it because it's only covering LetsEncrypt isn't valid, if anyone else wants to contribute alternatives that'd be welcomed, but it shouldn't be mandated if the information would be helpful.

Do you think it would be better suited and of value to the TLS docs page?

---

> I found it odd that you'd need to unfold them, because this is the content everyone is there for...

I can't really comment for this docs page as I am not familiar with what and how much of the information presented is relevant to all viewers on that page.

As stated earlier, there is value in collapsible admonitions when appropriate imo.

Regular admonitions still help with the layout of the page content and conveying the info at a glance. I would personally prefer using them but unless anyone is going to weigh in, either way is fine. The information documented is more important, presentation is complimentary to that.

---

> Why do you figure? I must say I disagree.

I find:

> A `Service` is required for getting the traffic to the pod itself. The service is somewhat crucial. Its configuration determines whether the original IP from the sender will be kept. [More about this further down below](https://pullrequest-2111--dms-doc-previews.netlify.app/config/advanced/kubernetes/#exposing-your-mail-server-to-the-outside-world). The configuration you're seeing does keep the original IP, but you will not be able to scale this way. We have chosen to go this route in this case because we think most K8s users will only want to have one instance anyway, and users that need high availability know how to do it anyways. You will want to have your load-balancer give this service an external, routable IP address.

to be worse UX vs:


> ### Service
> 
> A `Service` is required for getting the traffic to the pod itself. The service is somewhat crucial. Its configuration determines whether the original IP from the sender will be kept. [More about this further down below](https://pullrequest-2111--dms-doc-previews.netlify.app/config/advanced/kubernetes/#exposing-your-mail-server-to-the-outside-world).
> 
> The configuration you're seeing does keep the original IP, but you will not be able to scale this way. We have chosen to go this route in this case because we think most K8s users will only want to have one instance anyway, and users that need high availability know how to do it anyways.
> 
> You will want to have your load-balancer give this service an external, routable IP address.

which is easier to digest, especially in the context of the whole document.

There's already some verbosity from the writing style we could trim, but just giving the content some more breathing space really helps (_at least for me it's heavily appreciated, imagine my verbose responses compacted into just a few paragraphs.._).

If I have some k8s familiarity and have a setup already or looking back at these docs to reference something, is it possible that I might want to quickly lookup one of these manifests as reference? With the subheading it'll be added to the ToC sidebar under the mainfests heading level, and it'll be easier to scroll through the document and locate (_if not using the ToC_).

georglauterbach commented:
About the TLS / certificates topic: Kube-Lego is deprecated, and should not be used in new deployments anymore. You'd want to use CertManager. However, there are many possibilities. That's why I'd leave this section (or mention it in the TLS section seperately).

I have added more subsections to structure the document a bit more. I would however very much try to avoid putting the configuration files in admonitions that one has to unfold. With recent versions of MkDocs (which we are using), there is a dedicated copy button in the top right hand corner one can use to copy the code as a whole :)

I have also split the description for the service.

polarathene commented on docs/content/config/advanced/kubernetes.md:
```diff
@@ -460,79 +444,19 @@ Then, configure both [Postfix][docs-postfix] and [Dovecot][docs-dovecot] to expe
+- it is not possible to access the mailserver via cluster-DNS, as the PROXY protocol is required for incoming connections.
 
+[Kustomize]: https://kustomize.io/
+[TLS]: ../security/ssl.md
```
Convention for doc links is using the `docs-` prefix. I often keep it all lowercase, the idea with the prefix is that it'd be easy to parse docs with tooling or CLI utilities for use cases like identifying/checking internal doc links :man_shrugging: 

Maybe it's over thinking, and isn't necessary.

I wanted to add it to our CI at one point so our docs wouldn't miss changes breaking links. The Mkdocs build --strict feature generally works, but I do recall it having some gotchas such as not catching invalid anchor (`#)` links to jump to specific sections of our docs.

georglauterbach replied:
Done with af58f352188a16c0e12bab0ec033e7eb56838531 :)

georglauterbach commented:
I will test this ASAP, but I have to figure out how solve a problem with [Cilium](https://cilium.io/) in my cluster first. I will mark this as ready for review as soon as I know my mail server setup works.

polarathene commented on docs/content/config/advanced/kubernetes.md:
```diff
@@ -448,8 +448,8 @@ With this approach,
 
-[Kustomize]: https://kustomize.io/
-[TLS]: ../security/ssl.md
+[docs-kustomize]: https://kustomize.io/
```
I think there was a misunderstanding here, `kustomize` is an external link, not internal docs.

Usually that distinction is clear as I placed all the internal `docs-` links above the others. For example below is `dockerhub-haproxy`. You can continue to use `kustomize`, but `docs-` is misleading here :sweat_smile:

georglauterbach replied:
I will change this, but I need to adjust the link above so I cannot directly commit your suggested changes.

polarathene commented on docs/content/config/advanced/kubernetes.md:
```diff
@@ -2,392 +2,375 @@
 
-!!! warning
-    Any sensitive data (keys, etc) should be deployed via [Secrets][k8s-config-secret]. Other configuration just fits well into [ConfigMaps][k8s-config-pod].
+We can also make use of user-provided configuration files, e.g. `user-patches.sh`, `postfix-accounts.cf` and more, to adjust `docker-mailserver` to our likings. We encourage you to have a look at [Kustomize][docs-kustomize] for creating `ConfigMap`s from multiple files, but for now, we will provide a simple, hand-written example. This example is absolutely minimal and only goes to show what can b ...
```
```suggestion
We can also make use of user-provided configuration files, e.g. `user-patches.sh`, `postfix-accounts.cf` and more, to adjust `docker-mailserver` to our likings. We encourage you to have a look at [Kustomize][kustomize] for creating `ConfigMap`s from multiple files, but for now, we will provide a simple, hand-written example. This example is absolutely minimal and only goes to show what can be done.
```

Or if you prefer the `[Kustomize]` alone is fine with me too.

georglauterbach replied:
I just now saw this, alright, I can commit your changes :D

polarathene reviewed (approved):
Ideally someone with k8s knowledge can also chime in, but the rest seems good :+1:

georglauterbach commented:
@polarathene There will be Orr commits because I will remove small parts (I overlooked a few things when copy-wasting my setup). I will reach back to you and the others when this is ready again :)

radicand commented:
I'll try to review when I get some time

georglauterbach commented:
Ready for review.

georglauterbach commented:
@polarathene Sorry for dismissing your review again. Because of #2113 I realised I forgot to update the caps list :)

georglauterbach commented:
@radicand Have you had time to get a look at the updates?:)

radicand commented:
> @radicand Have you had time to get a look at the updates?:) If you don't have the time, I think it's not a big deal, I'd just need to know :D

Starting now...

radicand commented on docs/content/config/advanced/kubernetes.md:
```diff
@@ -2,392 +2,374 @@
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 1Gi
```
I'd suggest we either change this default to something larger, or make it explicit (as in providing guidance) on setting it. While PVCs can be resized, it is not _always easy_ to do based on the storage backend, and running out of space prematurely sours the experience for sure.

radicand commented on docs/content/config/advanced/kubernetes.md:
````diff
@@ -2,392 +2,374 @@
-            mountPath: /tmp/docker-mailserver/postfix-main.cf
-            readOnly: true
-    ```
+With this setup, you want your load-balancer (e.g. MetalLB) to give this service an external, especially routable IP address.
````
You do not need MetalLB to give you an external address - an internal address that you DNAT/port forward will do (this is how I run it actually). Since there is a later section talking about MetalLB anyway, I'd strike the line here and just reference that real client IP is discussed below.

radicand commented on docs/content/config/advanced/kubernetes.md:
```diff
@@ -2,392 +2,374 @@
+### Deployments
 
-!!! example
+Last but not least, the `Deployment` becomes the most complex component. The deployment brings maximal security measures without compromising on ease of use.
```
I find this a somewhat confusing sentence -- I'd suggest we just mention what the purpose of the Deployment object is rather than talking about security (which you can do, but a deployment's purpose is not to enforce security). Something like "Lastly, the Deployment is the most complex component. It instructs Kubernetes how to run the docker-mailserver container and how to apply your ConfigMaps and persisted storage. Additionally, we can set options to enforce runtime security here."

georglauterbach replied:
I find your sentence better too. I will copy it :)

radicand commented on docs/content/config/advanced/kubernetes.md:
```diff
@@ -2,392 +2,374 @@
+            seccompProfile:
+              type: RuntimeDefault
+
+          resources:
+            limits:
+              memory: 4Gi
+              cpu: 1500m
+            requests:
+              memory: 2Gi
+              cpu: 600m
```
We might want to add comments that these values can/should be tuned based on the scale of the implementation. I run the below for 2 (real) users, ClamAV off, and ~400 emails a day. Spam sucks.
```
            requests:
              memory: "500Mi"
              cpu: "750m"
            limits:
              memory: "1500Mi"
              cpu: "900m"
```

I mention this because K8s will start refusing to schedule pods if you exceed your memory/cpu budgets, and you don't _always_ need the suggested amounts. Users should be aware that they can tune this.

radicand commented on docs/content/config/advanced/kubernetes.md:
```diff
@@ -2,392 +2,374 @@
+            - name: transfer
+              containerPort: 25
+              protocol: TCP
+            - name: esmtp-implicit
+              containerPort: 465
+              protocol: TCP
```
This may be debatable, but 465 isn't a supported standard from what I'm aware, and nothing requires it (famous last words). Should we be good citizens and not advertise opening it up by removing it from here and the service?

polarathene replied:
465 for implicit TLS is valid, it has a bit of a rough history being deprecated from STMPS (_port 25 with implicit TLS_) to being designated as implicit TLS `submissions` service (_as opposed to explicit `submission` service with StartTLS on port 587_).

That change requires auth like port 587, but plain TLS socket can be used, no StartTLS negotiation required. Due to the confusion from it's history, it's support is a bit spotty when offered by services.

587 had been around for a while and established itself, so is more common to see and without any confusion. That said StartTLS has also had some issues with how some software behaved (_eg IIRC.. sending auth credentials in plain-text if StartTLS failed to negotiate a secure exchange and the server rejected the client)).

georglauterbach replied:
It's funny to see this debate everywhere :D @radicand you may want to read [this](https://github.com/authelia/authelia/issues/2194), and what @polarathene said is absolutely correct :)

TL;DR 465 is actually preferred over 578

radicand commented on docs/content/config/advanced/kubernetes.md:
```diff
@@ -2,392 +2,374 @@
+
+!!! attention "Sensitive Data"
+
+    For storing OpenDKIM keys, TLS certificates or any sort of sensitive data, you should be using `Secret`s. You can mount secrets like `ConfigMaps` and use them the same way.
```
Just pointing out a typo in removing the `s` from ConfigMaps like is done in Secrets just prior.

georglauterbach replied:
Done.

radicand commented on docs/content/config/advanced/kubernetes.md:
```diff
@@ -2,392 +2,374 @@
+
+    For storing OpenDKIM keys, TLS certificates or any sort of sensitive data, you should be using `Secret`s. You can mount secrets like `ConfigMaps` and use them the same way.
+
+## Exposing your Mailserver to the Outside World
```
For this section, I'd suggest we start with a recommendation, which is to use MetalLB and route to the LB IP via DNAT, direct routing, etc. If MetalLB can't be used, then here is our next best suggestion, but its caveats are...

This doesn't have to be changed, but as I mentioned above, an externally routable IP is not required for SPF to work when using a LB like MetalLB. I think we're just providing an impression that users need something that is not often available (a dedicated external IP they can use for this), when it's not really the case.

georglauterbach replied:
I will see what I can do, but I like the idea :)

radicand commented:
I'm done with my review, feel free to take or leave any of my points :)

georglauterbach commented:
Alright. Thank you @radicand for your suggestions :D

@polarathene I have enabled auto-merge. If you think this is okay, this PR should get merged when you approve it :)

radicand reviewed (approved):
LGTM

georglauterbach commented:
@polarathene You have already approved this PR a thousand times. Therefore, I'm going to assume you're fine with it and I'm beating you to the punch (a strange English phrase) with merging this PR so you don't have to approve it again :D
